### PR TITLE
Optimized the neighbor joining (NJ) tree building algorithm

### PR DESCRIPTION
**Pull request by qiyunzhu**
This PR optimizes [**neighbor joining (NJ)**](https://en.wikipedia.org/wiki/Neighbor_joining), a classical algorithm for phylogenetic tree building based on a distance matrix among taxa. The new algorithm is 20-30x faster than the old one, and is competitive when compared with popular implementations in the community.

FYI @christapo @RaeedA 

The old implementation was already quite efficient, due to the use of NumPy vectorization throughout the algorithm. The new algorithm further achieves performance gain because of:

1. The agglomerative clustering process does not create intermediate matrices and arrays during each iteration. Instead, it re-uses the same block of memory space over and over. This not only improves memory efficiency but also reduces the time overheads of allocating new memory space.
2. One time-consuming step: Building a "Q-matrix" and finding the minimum value, is Cythonized. It is possible to use NumPy to find the minimum non-diagonal value in a distance matrix, but the process is quite convoluted. The Cython code is more efficient in this.
3. Instead of building a Newick string incrementally during iteration, the new code builds a linkage matrix, which is purely numeric and can be pre-allocated. The linkage matrix data structure resembles [that of SciPy's](https://docs.scipy.org/doc/scipy/reference/generated/scipy.cluster.hierarchy.linkage.html), but with some differences to reflect the unequal branch lengths of the left and right children.
5. Multiple other tweaks of the math.

Note: The entire function works with numbers only, and can be completely Cythonized. However, Cythonization does not deliver much performance gain in the test (roughly 5-20%), likely because the code already uses NumPy in every line. But the code may be further optimized given its independence from any Pythonic operations.

Another potentional major optimization is to directly manipulate the input distance matrix rather than making a copy of it. This will be useful when working with extremely large distance matrices. However, this will destroy the input distance matrix, therefore requires more thoughts on the `DistanceMatrix` data structure before implementing.

***

Benchmarks on a distance matrix of **485** taxa, derived from real-world 16S rRNA sequences.

- scikit-bio (old): runtime = 0.441 sec, memory = 12,580 KB
- scikit-bio (new): runtime = 0.0161 sec, memory = 2,220 KB

Additional benchmarks were performed on Xeon Platinum 8358 CPUs:

485 taxa:

| Program | Runtime (sec) |
| --- | ---: |
| scikit-bio (new) | 0.03833 |
| scikit-bio (old) | 0.76472 |
| DendroPy 5.0.1 | 3.92156 |
| Biotite 1.0.1 | 0.13615 |
| BioPython 1.84 | 43.41698 |
| QuickTree 2.5 | 0.07855 |
| RapidNJ 2.3.3 | 0.05281 |

9232 taxa:

| Program | Runtime (sec) |
| --- | ---: |
| scikit-bio (new) | 140.638 |
| scikit-bio (old) | 7409.659 |
| Biotite 1.0.1 | 1209.335 |
| QuickTree 2.5 | 256.255 |
| RapidNJ 2.3.3 | 10.226 |

Notes:

- Runtime for Python libraries scikit-bio, DendroPy, Biotite and BioPython only includes the execution of the NJ algorithm, whereas runtime for command-line tools QuickTree and RapidNJ also include file input / output. Therefore, the benchmark is unfair to the latter category, especially the smaller test. scikit-bio isn't fast in reading a big distance matrix from a file. This needs further improvement.
- If I understand correctly, RapidNJ is heuristic, wheras others are optimal (NJ itself is heuristic for minimum evolution, but it is another story).
- I couldn't figure out how to install DecentTree, which is a very recent and performant NJ package.

***

Please complete the following checklist:

* [x] I have read the [contribution guidelines](https://scikit.bio/contribute.html).

* [x] I have documented all public-facing changes in the [changelog](https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/scikit-bio/scikit-bio/blob/main/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied.

  - **It is your responsibility to disclose** code, documentation, or other content derived from external source(s). If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] This pull request does not include code, documentation, or other content derived from external source(s).

Note: [This document](https://scikit.bio/devdoc/review.html) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

### CODE REVIEW

File: skbio/tree/_nj.py [hunk 1]
```diff
@@ -86,209 +112,232 @@ def nj(dm, disallow_negative_branch_length=True, result_constructor=None):
...
-    return i_to_u, j_to_u
+        taxon1, taxon2, length1, length2
+
+    """
+    # allocate node list
+    nodes = [TreeNode(name=x) for x in taxa] + [TreeNode() for _ in range(len(lm))]
+
+    # build tree incrementally
+    idx = len(taxa)
+    for c1, c2, l1, l2 in lm if rooted else lm[:-2]:
```
Comment (mataton) [inline skbio/tree/_nj.py:319]: `l1`, `l2` are hard to read in my opinion. They look very close to the numbers `11` and `12`. I'd suggest naming the variables something more descriptive or that avoids the use of a lot of lowercase `l`'s. A simple change to capital `L` would also suffice.

File: skbio/tree/_nj.py [hunk 2]
```diff
@@ -86,209 +112,232 @@ def nj(dm, disallow_negative_branch_length=True, result_constructor=None):
...
+
+    # final treatment of an unroot tree
+    if not rooted:
+        # this code assumes that the first element of the last row is the node
+        # see the end of _nj
+        c0, _, l0, _ = lm[-1]
+        c1, c2, l1, l2 = lm[-2]
+        c0, c1, c2 = nodes[int(c0)], nodes[int(c1)], nodes[int(c2)]
+        if clip_to_zero:
+            c0.length = l0 if l0 >= 0 else 0.0
```
Comment (mataton) [inline skbio/tree/_nj.py:337]: Same here, regarding readability issues of lowercase `l`.

**Review by mataton — CHANGES_REQUESTED** [review]
Besides the minor readability issues, it all looks good to me!

**Comment by qiyunzhu** [comment]
@mataton Thanks! I have addressed your comments.

**Pull request merged**